Select typeahead shows selected value instead of text on some renders

Hi, I'm experiencing an issue: with a `Select` of the `typeahead` (single) variant, the first render after making a selection displays the item's `value`, not its textContent, as the "typeaheadValue" inside the input. However, the *second* render replaces the value with the proper text.

From the user's perspective, it looks like this:
1. Click on the select's arrow to see the available options
2. Click on one of the options to select it → a different text (the **value**) appears as the selection in the input at the top
3. Click outside the select to close it → the proper text (the textContent between `<SelectOption>` and `</SelectOption>`) appears as the selection - this is what the user expects from the beginning
4. Alternatively, clicking the same option twice has the same result - on the second render, the option's text, not value, is rendered

I think it has something to do with this fix: https://github.com/patternfly/patternfly-react/pull/3096  
The linked change appears to "fix" setting the `selections` from outside by using the `selections` value as the type-ahead text. However, this seems to assume that the `value` and the displayed option text are always the same, which looks like an unreasonable assumption to make.

**Please provide the steps to reproduce. Feel free to link CodeSandbox or another tool.**
Here's a minimal example that shows the weird behavior: https://codesandbox.io/s/unruffled-chandrasekhar-idl9j?file=/src/App.tsx

**Is this a bug or enhancement? If this issue is a bug, is this issue blocking you or is there a work-around?**
This is a bug that has no work-around that I know of, as having a different option `value` and option text is absolutely necessary (values are often not human-readable).

Is there any other way to make this work? I've tried not setting `selections` and letting the Select component manage this internally, but this results in the input not updating at all (i.e. no user feedback as to which option is selected).

**What is your product and what release version are you targeting?**
Our product is an admin-type GUI and we're on `react-core` 3.153.3, though this is reproducible on newer versions also (as seen in the sample code linked above).

Hi @rkaw92 and @knokit 

I believe I figured out what is happening.

### Background

Select stores `TypeaheadInputValue` in its states to keep track on what option is hovered when you navigate with arrow keys or mouse.
The `selections` props is the one that handles what option/s was/were selected. 

### The flow:

When an option is selected in the first time, `props.selections` will contain the value (e.g. `FL`) and not the label `Florida` because that was is set by the user. 
Due to the change in the props, `componentDidUpdate` is being called [and changes `typeaheadInputValue` to the new `props.selections` - `FL`](https://github.com/patternfly/patternfly-react/blob/master/packages/react-core/src/components/Select/Select.tsx#L205-L209).
When we render the value that is displayed in the select box we check that `typeaheadInputValue` is not `null`. If it is `null` we show the current value using [`getDisplay`](https://github.com/patternfly/patternfly-react/blob/master/packages/react-core/src/components/Select/Select.tsx#L849-L853) otherwise we show `typeaheadInputValue`.
When we select again the same option from the menu, the menu is closed which [resets the `typeaheadInputValue` to `null`](https://github.com/patternfly/patternfly-react/blob/master/packages/react-core/src/components/Select/Select.tsx#L239-L247) and even though `componentDidUpdate` is called again this time because `props.selections` was not changed, `typeaheadInputValue` is not set and is `null` which later display the correct value.

To me, it seems like, the main reason for this problem is the checking in `componentDidUpdate`. Looking in the history, its purpose was [to update the selection without interacting with the menu (f.e. through `setTimeout`)](https://github.com/patternfly/patternfly-react/pull/3096). However, it looks like it was fixed in a better way and not by updating the `typeaheadInputValue`.

All in all, I removed that checking and verified that @SpyTec sandbox is [working even now](https://codesandbox.io/s/silly-keldysh-osdsj?file=/src/index.js). In addition, ran unit tests to make sure this doesn't break anything else and fixes the problem represented here.

I tried to run integration tests but the demo-apps is failing 
```
Html Webpack Plugin:

  Error: Child compilation failed:
  Module build failed (from /home/bshuster/Documents/projects/patternfly-react/n  ode_modules/react-scripts/node_modules/babel-loader/lib/index.js):
  Error: [BABEL] /home/bshuster/Documents/projects/patternfly-react/node_modules  /react-scripts/node_modules/webpack/buildin/global.js: Cannot find module '@ba  bel/helper-create-regexp-features-plugin'
  Require stack:
  - /home/bshuster/Documents/projects/patternfly-react/node_modules/babel-preset  -react-app/node_modules/@babel/plugin-transform-dotall-regex/lib/index.js
  - /home/bshuster/Documents/projects/patternfly-react/node_modules/babel-preset  -react-app/node_modules/@babel/preset-env/lib/available-plugins.js
  - /home/bshuster/Documents/projects/patternfly-react/node_modules/babel-preset  -react-app/node_modules/@babel/preset-env/lib/index.js
  - /home/bshuster/Documents/projects/patternfly-react/node_modules/babel-preset  -react-app/dependencies.js
  - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-script  s/node_modules/@babel/core/lib/config/files/plugins.js
  - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-script  s/node_modules/@babel/core/lib/config/files/index.js
  - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-script  s/node_modules/@babel/core/lib/index.js
  - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-script  s/node_modules/babel-loader/lib/index.js
  - /home/bshuster/Documents/projects/patternfly-react/node_modules/loader-runne  r/lib/loadLoader.js
  - /home/bshuster/Documents/projects/patternfly-react/node_modules/loader-runne  r/lib/LoaderRunner.js
  - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-script  s/node_modules/webpack/lib/NormalModule.js
  - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-script  s/node_modules/webpack/lib/NormalModuleFactory.js
  - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-script  s/node_modules/webpack/lib/Compiler.js
  - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-script  s/node_modules/webpack/lib/webpack.js
  - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-script  s/scripts/start.js (While processing: "/home/bshuster/Documents/projects/patte  rnfly-react/node_modules/babel-preset-react-app/dependencies.js")
  
  - loader.js:965 Function.Module._resolveFilename
    internal/modules/cjs/loader.js:965:15
  
  - loader.js:841 Function.Module._load
    internal/modules/cjs/loader.js:841:27
  
  - loader.js:1025 Module.require
    internal/modules/cjs/loader.js:1025:19
  
  - helpers.js:72 require
    internal/modules/cjs/helpers.js:72:18
  
  - index.js:8 Object.
    [patternfly-react]/[babel-preset-react-app]/[@babel]/plugin-transform-dotall    -regex/lib/index.js:8:41
  
  - loader.js:1137 Module._compile
    internal/modules/cjs/loader.js:1137:30
  
  - loader.js:1157 Object.Module._extensions..js
    internal/modules/cjs/loader.js:1157:10
  
  - loader.js:985 Module.load
    internal/modules/cjs/loader.js:985:32
  
  - loader.js:878 Function.Module._load
    internal/modules/cjs/loader.js:878:14
  
  - loader.js:1025 Module.require
    internal/modules/cjs/loader.js:1025:19
  
  - helpers.js:72 require
    internal/modules/cjs/helpers.js:72:18
  
  - available-plugins.js:21 Object.
    [patternfly-react]/[babel-preset-react-app]/[@babel]/preset-env/lib/availabl    e-plugins.js:21:29
  
  - loader.js:1137 Module._compile
    internal/modules/cjs/loader.js:1137:30
  
  - loader.js:1157 Object.Module._extensions..js
    internal/modules/cjs/loader.js:1157:10
  
  - loader.js:985 Module.load
    internal/modules/cjs/loader.js:985:32
  
  
  - Error: [BABEL] /home/bshuster/Documents/projects/patternfly-react/node_modul    es/react-scripts/node_modules/webpack/buildin/global.js: Cannot find module     '@babel/helper-create-regexp-features-plugin'
  
  - Require stack:
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/babel-pres    et-react-app/node_modules/@babel/plugin-transform-dotall-regex/lib/index.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/babel-pres    et-react-app/node_modules/@babel/preset-env/lib/available-plugins.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/babel-pres    et-react-app/node_modules/@babel/preset-env/lib/index.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/babel-pres    et-react-app/dependencies.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-scri    pts/node_modules/@babel/core/lib/config/files/plugins.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-scri    pts/node_modules/@babel/core/lib/config/files/index.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-scri    pts/node_modules/@babel/core/lib/index.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-scri    pts/node_modules/babel-loader/lib/index.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/loader-run    ner/lib/loadLoader.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/loader-run    ner/lib/LoaderRunner.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-scri    pts/node_modules/webpack/lib/NormalModule.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-scri    pts/node_modules/webpack/lib/NormalModuleFactory.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-scri    pts/node_modules/webpack/lib/Compiler.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-scri    pts/node_modules/webpack/lib/webpack.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-scri    pts/scripts/start.js (While processing: "/home/bshuster/Documents/projects/p    atternfly-react/node_modules/babel-preset-react-app/dependencies.js")
  
  - Module build failed (from /home/bshuster/Documents/projects/patternfly-react    /node_modules/react-scripts/node_modules/babel-loader/lib/index.js):
  
  - Error: [BABEL] /home/bshuster/Documents/projects/patternfly-react/node_modul    es/react-scripts/node_modules/webpack/buildin/module.js: Cannot find module     '@babel/helper-create-regexp-features-plugin'
  
  - Require stack:
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/babel-pres    et-react-app/node_modules/@babel/plugin-transform-dotall-regex/lib/index.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/babel-pres    et-react-app/node_modules/@babel/preset-env/lib/available-plugins.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/babel-pres    et-react-app/node_modules/@babel/preset-env/lib/index.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/babel-pres    et-react-app/dependencies.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-scri    pts/node_modules/@babel/core/lib/config/files/plugins.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-scri    pts/node_modules/@babel/core/lib/config/files/index.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-scri    pts/node_modules/@babel/core/lib/index.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-scri    pts/node_modules/babel-loader/lib/index.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/loader-run    ner/lib/loadLoader.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/loader-run    ner/lib/LoaderRunner.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-scri    pts/node_modules/webpack/lib/NormalModule.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-scri    pts/node_modules/webpack/lib/NormalModuleFactory.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-scri    pts/node_modules/webpack/lib/Compiler.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-scri    pts/node_modules/webpack/lib/webpack.js
  
  - - /home/bshuster/Documents/projects/patternfly-react/node_modules/react-scri    pts/scripts/start.js (While processing: "/home/bshuster/Documents/projects/p    atternfly-react/node_modules/babel-preset-react-app/dependencies.js")
  

```

--- a/packages/react-core/src/components/Select/Select.tsx
+++ b/packages/react-core/src/components/Select/Select.tsx
@@ -202,12 +202,6 @@ export class Select extends React.Component<SelectProps & OUIAProps, SelectState
       });
     }
 
-    if (prevProps.selections !== this.props.selections && this.props.variant === SelectVariant.typeahead) {
-      this.setState({
-        typeaheadInputValue: this.props.selections as string
-      });
-    }
-
     if (
       this.props.onFavorite &&
       (this.props.favorites.length !== prevProps.favorites.length ||

--- a/packages/react-integration/cypress/integration/select.spec.ts
+++ b/packages/react-integration/cypress/integration/select.spec.ts
@@ -19,6 +19,23 @@ describe('Select Test', () => {
       .should('exist');
   });
 
+  it('Verify Typeahead Select with labels', () => {
+    const find = (selector: string) =>
+      cy
+        .get('#typeahead-select-labels-id')
+        .parent()
+        .find(selector);
+    find('button.pf-c-select__toggle-button').click();
+    find('li:nth-child(3) button').click();
+    find('#typeahead-select-label-select-typeahead').should('have.value', 'Florida');
+    find('button.pf-c-select__toggle-clear:first').click();
+    find('#typeahead-select-label-select-typeahead').should('have.value', '');
+    cy.get('#button-typeahead-labels').click();
+    find('#typeahead-select-label-select-typeahead').should('have.value', 'New York');
+    find('button.pf-c-select__toggle-clear:first').click();
+    find('#typeahead-select-label-select-typeahead').should('have.value', '');
+  });
+
   it('Verify Single Select', () => {
     cy.get('#single-select').click();
     cy.get('#Miss > .pf-c-select__menu-item').click();

--- a/packages/react-integration/demo-app-ts/src/components/demos/SelectDemo/SelectDemo.tsx
+++ b/packages/react-integration/demo-app-ts/src/components/demos/SelectDemo/SelectDemo.tsx
@@ -1,4 +1,5 @@
 import {
+  Button,
   Select,
   SelectOption,
   SelectVariant,
@@ -54,6 +55,8 @@ export interface SelectDemoState {
   noBadgeCheckSelected: string[];
   menuDocumentBodyisOpen: boolean;
   menuDocumentBodySelected: string[];
+  lbltypeaheadSelected: string;
+  lbltypeaheadisOpen: boolean;
 }
 
 export class SelectDemo extends Component<SelectDemoState> {
@@ -78,6 +81,8 @@ export class SelectDemo extends Component<SelectDemoState> {
     typeaheadMultiSelected: [''],
     cdtypeaheadMultiisOpen: false,
     cdtypeaheadMultiSelected: [] as string[],
+    lbltypeaheadSelected: '',
+    lbltypeaheadisOpen: false,
     plainTypeaheadMultiisOpen: false,
     plainTypeaheadMultiSelected: [''],
     plainTypeaheadMultiIsPlain: true,
@@ -90,6 +95,13 @@ export class SelectDemo extends Component<SelectDemoState> {
       { value: 'Florida', disabled: false },
       { value: 'New Jersey', disabled: false },
       { value: 'Texas', disabled: false }
+    ],
+    typeaheadLabelOptions: [
+      { label: 'Alabama', value: 'AL', disabled: false },
+      { label: 'New York', value: 'NY', disabled: false },
+      { label: 'Florida', value: 'FL', disabled: false },
+      { label: 'New Jersey', value: 'NJ', disabled: false },
+      { label: 'North Carolina', value: 'NC', disabled: false }
     ],
     typeaheadIsCreatable: false,
     typeaheadNewOptions: false,
@@ -228,6 +240,10 @@ export class SelectDemo extends Component<SelectDemoState> {
     this.setState({
       cdtypeaheadMultiisOpen
     });
+  };
+
+  lbtypaheadToggle = (lbltypeaheadisOpen: boolean) => {
+    this.setState({ lbltypeaheadisOpen });
   };
 
   plainTypeaheadMultiOnToggle = (plainTypeaheadMultiisOpen: boolean) => {
@@ -433,6 +449,24 @@ export class SelectDemo extends Component<SelectDemoState> {
     }
   };
 
+  lbltypeaheadSelect = (
+    _event: React.MouseEvent | React.ChangeEvent,
+    selection: string | SelectOptionObject,
+    isPlaceholder?: boolean
+  ) => {
+    if (isPlaceholder) {
+      this.clearSelection();
+      return;
+    }
+    this.setState(
+      {
+        lbltypeaheadSelected: selection,
+        lbltypeaheadisOpen: false
+      },
+      () => console.log('selections: ', this.state.lbltypeaheadSelected)
+    );
+  };
+
   plainTypeaheadMultiOnSelect = (
     event: React.MouseEvent | React.ChangeEvent,
     selection: string | SelectOptionObject
@@ -501,6 +535,8 @@ export class SelectDemo extends Component<SelectDemoState> {
       typeaheadMultiisOpen: false,
       cdtypeaheadMultiisOpen: false,
       cdtypeaheadMultiSelected: [],
+      lbltypeaheadSelected: '',
+      lbltypeaheadisOpen: false,
       plainTypeaheadMultiSelected: [''],
       plainTypeaheadMultiisOpen: false,
       customTypeaheadMultiSelected: [''],
@@ -848,6 +884,47 @@ export class SelectDemo extends Component<SelectDemoState> {
     );
   }
 
+  renderLabelTypeaheadSelect() {
+    const { typeaheadLabelOptions, lbltypeaheadisOpen, lbltypeaheadSelected } = this.state;
+    const titleId = 'typeahead-select-labels-id';
+    return (
+      <StackItem isFilled={false}>
+        <Title headingLevel="h2" size="2xl">
+          Typeahead Select with labelled options
+        </Title>
+        <div>
+          <span id={titleId} hidden>
+            Select a state
+          </span>
+          <Select
+            toggleId="typeahead-select-label"
+            variant={SelectVariant.typeahead}
+            aria-label="Select a state"
+            onToggle={this.lbtypaheadToggle}
+            onSelect={this.lbltypeaheadSelect}
+            onClear={this.clearSelection}
+            selections={lbltypeaheadSelected}
+            isOpen={lbltypeaheadisOpen}
+            aria-labelledby={titleId}
+            placeholderText="Select a state"
+          >
+            {typeaheadLabelOptions.map((option, index) => (
+              <SelectOption isDisabled={option.disabled} key={index} value={option.value} id={option.value}>
+                {option.label}
+              </SelectOption>
+            ))}
+          </Select>
+          <Button
+            id="button-typeahead-labels"
+            onClick={() => this.lbltypeaheadSelect(null, typeaheadLabelOptions[1].value)}
+          >
+            Select {typeaheadLabelOptions[1].label}
+          </Button>
+        </div>
+      </StackItem>
+    );
+  }
+
   renderTypeaheadMultiSelect() {
     const { typeaheadMultiisOpen, typeaheadMultiSelected } = this.state;
     const titleId = 'multi-typeahead-select-id';
@@ -1117,6 +1194,7 @@ export class SelectDemo extends Component<SelectDemoState> {
         {this.renderMenuOnDocumentBodySelect()}
         {this.renderDescriptionSelect()}
         {this.renderSelectWithDivider()}
+        {this.renderLabelTypeaheadSelect()}
       </Stack>
     );
   }

